### PR TITLE
RE2022-231: parse biolog excel file into heatmap

### PR DESCRIPTION
**Pull request by Tianhao-Gu**

### CODE REVIEW

File: src/loaders/genome_collection/parse_biolog_data.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        df = pd.read_excel(excel_file, sheet_name=sheet_name)
+
+        return df
+
+    except FileNotFoundError:
+        print(f"File '{excel_file}' not found.")
+
+
+def _read_biolog_data(biolog_data_file: Path) -> pd.DataFrame:
+    # Read the Biolog data file into a DataFrame and do some data cleaning
```
Comment (MrCreosote) [inline src/loaders/genome_collection/parse_biolog_data.py:50]: This method is really well commented, helps a lot

File: src/loaders/genome_collection/parse_biolog_data.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+def _retrieve_kbase_genome_id(meta_df, strain_id) -> str:
+    # Retrieve the KBase genome ID for the given strain ID from the metadata file (dataframe).
+
+    matching_row = meta_df[(meta_df['Strain ID'] == strain_id) & meta_df['Genome_Ref'].notna()]
+
+    if matching_row.empty:
+        raise ValueError(f"No matching Strain ID with a non-NaN Genome_Ref found for {strain_id}.")
+    else:
+        genome_ref = matching_row['Genome_Ref'].values[0]
```
Comment (MrCreosote) [inline src/loaders/genome_collection/parse_biolog_data.py:103]: Aren't we using the assembly ref as the kbase ID everywhere else?

File: src/loaders/genome_collection/parse_biolog_data.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+    min_value, max_value = float('inf'), float('-inf')
+    for strain_id, row in data_df.iterrows():
+        try:
+            genome_ref = _retrieve_kbase_genome_id(meta_df, strain_id)
+        except ValueError as e:
+            # TODO: this is a temporary solution to handle the missing KBase genome ID for some strains.
+            # Two strains 'PDO1076' and 'PTD-1' are missing the KBase genome ID in the metadata file.
+            print(f'Cannot find KBase genome ID for strain ID: {strain_id}.')
+            genome_ref = f'kbase_id_not_found - {str(uuid.uuid4())}'
```
Comment (MrCreosote) [inline src/loaders/genome_collection/parse_biolog_data.py:154]: For now I would just skip them. The server should handle that ok, it expects that not every row in genome attribs will have heatmap data

File: src/loaders/genome_collection/parse_biolog_data.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    :param root_dir: the root directory for the data to be processed
+    """
+    data_df = _read_biolog_data(biolog_data_file)
+    meta_df = _read_biolog_meta(biolog_meta_file)
+
+    heatmap_cell_details, heatmap_rows, heatmap_meta_dict = list(), list(), dict()
+
+    # create document for heatmap meta
+    categories = list()
+    for idx, metabolite in enumerate(data_df.columns):
```
Comment (MrCreosote) [inline src/loaders/genome_collection/parse_biolog_data.py:132]: A lot of this code looks very similar to the parser code. Should it be moved to a shared module?

I was originally thinking this should just output data as if it were run by a tool container and have the parser summarize it but that might be overkill.

File: src/loaders/genome_collection/parse_biolog_data.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    # Drop unused rows ('total cpds. Utilized', 'strain designation' and unused rows at the end.)
+    data_df = data_df.drop(index=[0, 1, 194, 195, 196, 197, 198])
+
+    # Make the first column ('strain designation') the index
+    data_df = data_df.set_index('strain designation')
+
+    # Transpose the DataFrame
+    data_df.index.name = None
+    data_df.columns.name = None
```
Comment (MrCreosote) [inline src/loaders/genome_collection/parse_biolog_data.py:71]: Why are these lines necessary?

File: src/loaders/genome_collection/parse_biolog_data.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
+import uuid
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: I think we will only generate this once. I am planning just call `generate_pmi_biolog_heatmap_data` in a python shell.

File: src/loaders/genome_collection/parse_biolog_data.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+from pathlib import Path
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: 👍

File: src/loaders/genome_collection/parse_biolog_data.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+        return genome_ref
+
+
+def generate_biolog_heatmap_data(
+        biolog_data_file: Path,
+        biolog_meta_file: Path,
+        load_ver: str,
+        env: str = 'PROD',
+        kbase_collection: str = 'biolog',
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: updated.

File: src/loaders/genome_collection/parse_biolog_data.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+        root_dir: str = loader_common_names.ROOT_DIR):
+    """
+    Generate JSONL files with the heatmap data for the Biolog data file.
+
+    :param biolog_data_file: Biolog data file
+    :param biolog_meta_file: Biolog meta file
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: 👍

File: src/loaders/genome_collection/parse_biolog_data.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    # link to the file: https://docs.google.com/spreadsheets/d/1QmC6UHWOEVfpmrveRBl_izictbNeN1PA/edit#gid=1135979967
+
+    # Drop unused columns
+    data_df = data_df.drop(columns=['# pos. strains'])
+
+    # Make the first row ('strain designation') the column names
+    data_df.columns = data_df.loc[1]
+
+    # Drop unused rows ('total cpds. Utilized', 'strain designation' and unused rows at the end.)
+    data_df = data_df.drop(index=[0, 1, 194, 195, 196, 197, 198])
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: added more notes to explain. I made `strain designation` the column at previous step. But the row `strain designation` still exists. Therefore dropping again.

[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: added more notes to explain. It actually takes `Carbon Source` as column and then `total cpds. Utilized` as index 0.

File: src/loaders/genome_collection/parse_biolog_data.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+    data_df.index = data_df.index.str.replace(r'*', '').str.replace(r'**', '')
+
+    # Merge rows BT03 A(10/1) and BT03 B(11/18) into one row BT03
+    data_df = data_df.drop(index='BT03 B(11/18)')
+    data_df = data_df.rename(index={'BT03 A(10/1)': 'BT03'})
+
+    # Replace NaN with 0
+    # TODO: this is a temporary solution to handle the missing data for some strains.
+    # Two NaN values are present in the data file for 'YR530' and 'CF286' in the row labeled '3-Hydroxy 2-Butanone'.
+    data_df.fillna(0, inplace=True)
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: yes. There are 2 data entries with missing data.

File: src/loaders/genome_collection/parse_biolog_data.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+    return data_df
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: the media data? Which row/col are you referring to? How would we use that in the heatmap?

File: src/loaders/genome_collection/parse_biolog_data.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+    data_df.index.name = None
+    data_df.columns.name = None
+    data_df = data_df.transpose()
+
+    # Remove asterisks from index names
+    data_df.index = data_df.index.str.replace(r'*', '').str.replace(r'**', '')
+
+    # Merge rows BT03 A(10/1) and BT03 B(11/18) into one row BT03
+    data_df = data_df.drop(index='BT03 B(11/18)')
+    data_df = data_df.rename(index={'BT03 A(10/1)': 'BT03'})
```
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: 👍

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: noted. I will implement the mapping logic for CI. Now I just raise error if `env=='CI'`.

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 1 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: 👍

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 2 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: updated.

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 3 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: 👍

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 4 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: Yes, I considered that approach. However, for the microtrait dataset, we had separate folders with data for each row, which required processing and merging. In contrast, for this dataset, we already have all the necessary data, so I built all the models simultaneously. While there are similarities, the two approaches are also quite distinct.

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 5 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: removed. With or without, the result jsonl files are the same.

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: That's fine, but I wouldn't assume it'll only get run once. It'll need to be run for each environment at least, and we may need to regenerate the data at some point, and if the heatmap data structures change we'll need to re run it.

Either way, can you add an example execution to the *.md documentation?

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 4 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: Ok, my main concern here is that if the heatmap data structures change we'll need to update 2 places and we might forget. Can you add comments to this file and the parser pointing out the similar code in the other file and the need to update both if they change?

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: It's the last 3 lines of the original spreadsheet: 
```
Growth media:MOPS minimal media
* MOPS minimal media + 0.41uM Biotin
**MOPS minimal media + 0.41uM Biotin + 4mM Arginin + 0.68mM Glutamine
```
e.g. you'd figure out the media via the number of asterisks on the column. In the ticket I suggested it might be better to add a row to hold the media data rather than the asterisks, but up to you

The media data would go here: https://github.com/kbase/collections/blob/main/src/common/product_models/heatmap_common_models.py#L134-L137

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 1 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 5 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: Without a mapping, this code will only work for prod. Maybe it should throw an error if there's any other env?

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 3 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: completely missed that from ticket. Added logic to get the media description.

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: 👍

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 4 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: 👍

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 4 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_biolog_data.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: It looks like this will only set medias for columns with `*` and `**`, missing the `MOPS minimal media` columns without any asterisks

[the remark this answers is not in the document]
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_biolog_data.py]: I see. Added a default media for the rest of columns.

[the remark this answers is not in the document]
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_PMI_biolog_data.py [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+    return data_df
+
+
+def _read_biolog_meta(biolog_meta_file: Path) -> pd.DataFrame:
+    meta_df = _read_excel_as_df(biolog_meta_file)
+
+    return meta_df
+
+
+def _retrieve_kbase_aseembly_id(meta_df, strain_id) -> str:
```
Comment (MrCreosote) [inline src/loaders/genome_collection/parse_PMI_biolog_data.py:130]:
```suggestion
def _retrieve_kbase_assembly_id(meta_df, strain_id) -> str:
```

File: src/loaders/genome_collection/parse_PMI_biolog_data.py [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+def _read_excel_as_df(excel_file: Path, sheet_name=0) -> pd.DataFrame:
+    # Read the Excel file into a DataFrame
+
+    try:
+        df = pd.read_excel(excel_file, sheet_name=sheet_name)
+
+        return df
+
+    except FileNotFoundError:
+        print(f"File '{excel_file}' not found.")
```
Comment (MrCreosote) [inline src/loaders/genome_collection/parse_PMI_biolog_data.py:52]: Shouldn't this raise? Otherwise the code will continue with a null file

File: src/loaders/genome_collection/parse_PMI_biolog_data.py [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+
+    # Drop unused rows ('total cpds. Utilized', 'strain designation' and unused rows at the end.)
+    # NOTE: Drop 'strain designation' here because it also becomes the column names at the previous step.
+    #       Row 'Carbon Source' is used as the index upon reading the file.
+    #       Therefore, dataframe row index 0 starts at 'total cpds. Utilized' and
+    #       the last row index with data is 193 (row 195 '3-Hydroxy 2-Butanone' in the Excel sheet).
+    data_df = data_df.drop(index=[0, 1, 194, 195, 196, 197, 198])
+
+    # Make the first column ('strain designation') the index
+    data_df.loc[GROWTH_MEDIA_COL_NAME, STRAIN_DESIGNATION_COL_NAME] = GROWTH_MEDIA_COL_NAME
```
Comment (MrCreosote) [inline src/loaders/genome_collection/parse_PMI_biolog_data.py:103]: I'm not understanding what's going on here

File: src/loaders/genome_collection/parse_PMI_biolog_data.py (hunk 14 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_PMI_biolog_data.py]: 👍

File: src/loaders/genome_collection/parse_PMI_biolog_data.py (hunk 15 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_PMI_biolog_data.py]: 👍

File: src/loaders/genome_collection/parse_PMI_biolog_data.py (hunk 16 above)
Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_PMI_biolog_data.py:104]: added a comment to explain this. It was needed because I want to set `strain designation` as index in the next step. And currently, the cell was filled with the default media description, but we want the GROWTH_MEDIA_COL_NAME as column name.

File: src/loaders/genome_collection/parse_PMI_biolog_data.py (hunk 15 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_PMI_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_PMI_biolog_data.py (hunk 14 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_PMI_biolog_data.py]: :+1:

File: src/loaders/genome_collection/parse_PMI_biolog_data.py (hunk 16 above)
Comment (MrCreosote) [reply src/loaders/genome_collection/parse_PMI_biolog_data.py:104]: Maybe something like
```
Switch the value of the cell in row `growth_media` and column `strain designation` from the default media string to `growth_media`. This is needed so that when we make `strain designation` the index in the next step `growth_media` will be a row in that index.
```

I think I've got that right but feel free to improve

Comment (Tianhao-Gu) [reply src/loaders/genome_collection/parse_PMI_biolog_data.py:104]: 👍

Comment (MrCreosote) [reply src/loaders/genome_collection/parse_PMI_biolog_data.py:104]: :+1:

**Review by MrCreosote — APPROVED** [review]
LGTM. Like we mentioned, in the future we should probably turn this into a parser based on a cross-source standard and make smaller scripts to deal with ideosyncracies / differences in inputs from the source, or just work with the source to reformat their data (e.g. making some changes to the PMI sheet would make it much simpler to parse).
